# Task 2 - Qwen Dense Model

## Objective

Evaluate the execution of a Qwen Dense model in MaxText across different hardware backends and analyze platform-specific limitations.

The experiments focus on:

- Model discovery and configuration analysis
- CPU execution
- GPU execution
- TPU execution
- Comparison of platform behavior

---

# Qwen 0.6B - CPU Execution

## Objective

Evaluate the execution of the Qwen3-0.6B model on a CPU backend using MaxText.

This experiment serves as an initial smoke test to verify model configuration, dataset configuration, and training pipeline initialization.

---

## Environment

- Runtime: Google Colab CPU
- MaxText Version: Latest GitHub version
- JAX Version: 0.10.0
- Flax Version: 0.12.7

---

## Model Discovery

The MaxText repository was inspected to identify available Qwen model configurations.

Relevant configuration files discovered:

- `qwen3-0.6b.yml`
- `qwen3-1.7b.yml`
- `qwen3-4b.yml`
- `qwen3-8b.yml`
- `qwen3-14b.yml`
- Additional Qwen MoE and Next variants

The smallest dense model (`qwen3-0.6b`) was selected for initial experimentation.

---

## Configuration Analysis

The selected configuration defines:

- Embedding Dimension: 1024
- Query Heads: 16
- Key/Value Heads: 8
- Decoder Layers: 28
- MLP Dimension: 3072
- Vocabulary Size: 151,936
- Decoder Block: Qwen3
- Tokenizer Type: HuggingFace

---

## Environment Verification

The following components were verified successfully:

```python
import jax
import flax
import maxtext
```

Output:

```text
JAX: 0.10.0
Flax: 0.12.7
MaxText imported successfully
```

---

## Training Command

```bash
python3 -m maxtext.trainers.pre_train.train \
src/maxtext/configs/base.yml \
model_name=qwen3-0.6b \
dataset_type=synthetic \
run_name=qwen3_cpu_smoke_test \
steps=5 \
per_device_batch_size=1 \
enable_checkpointing=false
```

---

## Results

- MaxText initialized successfully.
- Qwen3-0.6B configuration loaded successfully.
- Synthetic dataset configuration was accepted.
- Training initialization started.

Execution failed before training could begin.

---

## Error Encountered

```text
ValueError:
Only interpret mode is supported on CPU backend.
```

---

## Observations

- Repository setup was successful.
- Model configuration was loaded correctly.
- The failure occurred during execution of Pallas-based attention kernels.
- These kernels are not supported on the CPU backend.

---

## Conclusion

The CPU execution attempt was unsuccessful.

The failure was caused by a backend limitation rather than a model configuration issue. The Qwen3-0.6B model requires execution paths that are not currently supported on the CPU backend used in this environment.

**Status: Failed due to CPU backend limitations.**

In [1]:
import jax
print(jax.devices())

[CpuDevice(id=0)]


In [1]:
!git clone https://github.com/AI-Hypercomputer/maxtext.git

Cloning into 'maxtext'...
remote: Enumerating objects: 97286, done.
remote: Counting objects: 100% (34/34), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 97286 (delta 8), reused 13 (delta 4), pack-reused 97252 (from 1)
Receiving objects: 100% (97286/97286), 410.23 MiB | 20.27 MiB/s, done.
Resolving deltas: 100% (71476/71476), done.


In [2]:
%cd /content/maxtext

/content/maxtext


In [3]:
%%bash
bash src/dependencies/scripts/setup.sh MODE=stable

Checking current directory...
Directory check passed.
Skipping editing needrestart.conf
Checking Python version...
Python version check passed. Continuing with script.
--------------------------------------------------
Get:1 https://cli.github.com/packages stable InRelease [3,917 B]
Get:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:5 https://cli.github.com/packages stable/main amd64 Packages [354 B]
Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Get:7 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Get:8 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]
Get:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease [18.1 kB]
Get:10 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [7,221 kB]
Get:11 http://security.ubuntu.com/

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)


W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)


debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 2.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0  

In [2]:
import jax
print(jax.devices())
print("JAX:", jax.__version__)

[CpuDevice(id=0)]
JAX: 0.10.0


In [3]:
import flax
print("Flax:", flax.__version__)

Flax: 0.12.7


In [4]:
import maxtext
print("MaxText imported successfully")

[transformers] DeepseekV32Config got `key=rope_scaling` in kwargs but hasn't set it as attribute. For RoPE standardization you need to set `self.rope_parameters` in model's config. 


MaxText imported successfully


In [19]:
!ls src/maxtext/trainers/pre_train

__init__.py  scripts  train_compile.py	train.py


In [20]:
!python -m maxtext.trainers.pre_train.train \
src/maxtext/configs/base.yml \
model_name=qwen3-0.6b \
dataset_type=synthetic \
run_name=qwen3_smoke_test \
base_output_directory=/tmp/maxtext_logs \
steps=5 \
per_device_batch_size=1 \
enable_checkpointing=false

[transformers] DeepseekV32Config got `key=rope_scaling` in kwargs but hasn't set it as attribute. For RoPE standardization you need to set `self.rope_parameters` in model's config. 
2026-06-16 15:40:24.020987: E external/local_xla/xla/stream_executor/cuda/cuda_platform.cc:51] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)
I0616 15:40:24.328004 18916161871872 max_utils.py:285] Attempting to initialize the jax distributed system for TPU backend...
I0616 15:40:24.426170 18916161871872 xla_bridge.py:822] Unable to initialize backend 'tpu': ABORTED: Internal error when accessing libtpu multi-process lockfile. Run "$ sudo rm /tmp/libtpu_lockfile".
I0616 15:40:24.428324 18916161871872 pyconfig.py:559] Config param abort_on_inf_loss: True
I0616 15:40:24.428452 18916161871872 pyconfig.py:559] Config param abort_on_nan_loss: True
I0616 15:40:24.428517 18916161871872 pyconfig.py:559] Config param act_quantization_calibration_method: absmax
I0616 15:40:24.4

In [21]:
!python -m maxtext.trainers.pre_train.train \
src/maxtext/configs/base.yml \
model_name=qwen3-0.6b \
dataset_type=synthetic \
run_name=qwen3_smoke_test \
base_output_directory=/tmp/maxtext_logs \
steps=5 \
per_device_batch_size=1 \
enable_checkpointing=false \
> qwen_smoke.log 2>&1

In [22]:
!python -c "import jax; print(jax.devices())"

[CpuDevice(id=0)]


Result:
CPU execution failed because Pallas attention kernels are not supported on CPU backend.

# Qwen 0.6B - GPU Execution

## Objective

Evaluate the execution of the Qwen3-0.6B model on a GPU backend using MaxText and compare its behavior with the CPU execution.

---

## Environment

- Runtime: Google Colab GPU
- GPU: Tesla T4
- MaxText Version: Latest GitHub version
- JAX Version: 0.10.0
- Flax Version: 0.12.7

---

## Expected Outcome

Unlike the CPU execution, which failed due to unsupported Pallas attention kernels, the GPU backend was expected to support these kernels and allow successful model execution.

---

## Steps Performed

1. Switched Colab runtime to GPU.
2. Verified GPU availability through JAX.
3. Reinstalled MaxText dependencies.
4. Verified JAX, Flax, and MaxText installation.
5. Executed a Qwen3-0.6B synthetic training smoke test.
6. Collected and analyzed execution logs.

---

## GPU Verification

```python
import jax
print(jax.devices())
print("JAX:", jax.__version__)
```

Output:

```text
[CudaDevice(id=0)]
JAX: 0.10.0
```

---

## Training Command

```bash
python3 -m maxtext.trainers.pre_train.train \
src/maxtext/configs/base.yml \
model_name=qwen3-0.6b \
dataset_type=synthetic \
run_name=qwen3_gpu_smoke_test \
base_output_directory=/tmp/maxtext_logs \
steps=5 \
per_device_batch_size=1 \
hardware=gpu \
enable_checkpointing=false
```

---

## Results

- MaxText initialized successfully.
- JAX distributed system initialized on GPU.
- Qwen3-0.6B configuration loaded successfully.
- Synthetic dataset configuration was accepted.
- Training initialization started successfully.

Execution failed before training could begin.

---

## Error Encountered

```text
ModuleNotFoundError: No module named 'transformer_engine.jax'

jax.errors.JaxRuntimeError:
FAILED_PRECONDITION:
Triton support is only enabled for Ampere GPUs
(compute capability 8.0) and up,
but got compute capability 7.5.
```

---

## Observations

- GPU detection worked correctly.
- MaxText environment setup was successful.
- The failure was not caused by configuration issues.
- The Tesla T4 GPU provided by Google Colab has compute capability 7.5.
- The Triton kernels required by this MaxText configuration require Ampere-generation GPUs (compute capability 8.0+) or newer.

---

## Conclusion

The GPU execution attempt was unsuccessful due to hardware limitations of the available Tesla T4 GPU.

Although the MaxText environment, Qwen configuration, and GPU initialization all worked correctly, the training workflow requires Triton kernels that are only supported on Ampere-generation GPUs and newer.

**Status: Failed due to GPU hardware compatibility limitations.**

In [2]:
import jax
print(jax.devices())
print("JAX:", jax.__version__)

[CudaDevice(id=0)]
JAX: 0.7.2


In [4]:
%%bash
cd /content/maxtext
bash src/dependencies/scripts/setup.sh MODE=stable DEVICE=gpu

Checking current directory...
Directory check passed.
Skipping editing needrestart.conf
Checking Python version...
Python version check passed. Continuing with script.
--------------------------------------------------
Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Get:2 https://cli.github.com/packages stable InRelease [3,917 B]
Get:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1,581 B]
Get:4 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ Packages [95.6 kB]
Get:5 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Get:6 https://cli.github.com/packages stable/main amd64 Packages [354 B]
Get:7 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Get:8 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [2,764 kB]
Get:9 http://archive.ubuntu.com/ubuntu jammy InRelease [270 kB]
Get:10 https://r2u.stat.illinois.edu/ubuntu jammy/main amd64 Pa

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
W: https://packages.cloud.google.com/apt/dists/gcsfuse-jammy/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.


W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
W: https://packages.cloud.google.com/apt/dists/gcsfuse-jammy/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:

In [1]:
import jax
print(jax.__version__)   # should be 0.10.x

import flax
print(flax.__version__)  # should be 0.12.7

import maxtext

0.10.0
0.12.7


[transformers] DeepseekV32Config got `key=rope_scaling` in kwargs but hasn't set it as attribute. For RoPE standardization you need to set `self.rope_parameters` in model's config. 


In [23]:
!find . -name "*qwen*"

./src/maxtext/experimental/agent/ckpt_conversion_agent/context/qwen3
./src/maxtext/experimental/agent/ckpt_conversion_agent/ground_truth/qwen3.py
./src/maxtext/configs/models/qwen3-0.6b.yml
./src/maxtext/configs/models/qwen3-1.7b.yml
./src/maxtext/configs/models/qwen2.5-14b.yml
./src/maxtext/configs/models/qwen3-1.7b-base.yml
./src/maxtext/configs/models/qwen3-32b.yml
./src/maxtext/configs/models/qwen3-4b.yml
./src/maxtext/configs/models/qwen2.5-7b.yml
./src/maxtext/configs/models/qwen3-4b-thinking-2507.yml
./src/maxtext/configs/models/qwen2.5-1.5b.yml
./src/maxtext/configs/models/qwen3-4b-base.yml
./src/maxtext/configs/models/qwen3.5-35b-a3b.yml
./src/maxtext/configs/models/qwen3-8b.yml
./src/maxtext/configs/models/qwen3-30b-a3b.yml
./src/maxtext/configs/models/qwen3-30b-a3b-base.yml
./src/maxtext/configs/models/qwen3-next-80b-a3b.yml
./src/maxtext/configs/models/qwen3.5-397b-a17b.yml
./src/maxtext/configs/models/qwen3-custom-30b-a3b.yml
./src/maxtext/configs/models/qwen3-480b-a35b.ym

In [24]:
!sed -n '1,200p' src/maxtext/configs/models/qwen3-0.6b.yml

# Copyright 2023–2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# model config for qwen3-0.6b

base_emb_dim: 1024
base_num_query_heads: 16
base_num_kv_heads: 8
base_mlp_dim: 3072
base_num_decoder_layers: 28
head_dim: 128
mlp_activations: ["silu", "linear"] # "hidden_act": "silu" implies SwiGLU
vocab_size: 151936

decoder_block: "qwen3"

normalization_layer_epsilon: 1.0e-6
rope_max_timescale: 1000000

use_qk_norm: true

logits_via_embedding: true # from "tie_word_embeddings": true

In [21]:
%%bash
cd /content/maxtext

python3 -m maxtext.trainers.pre_train.train \
src/maxtext/configs/base.yml \
model_name=qwen3-0.6b \
dataset_type=synthetic \
run_name=qwen3_gpu_smoke_test \
base_output_directory=/tmp/maxtext_logs \
steps=5 \
per_device_batch_size=1 \
hardware=gpu \
enable_checkpointing=false

[transformers] DeepseekV32Config got `key=rope_scaling` in kwargs but hasn't set it as attribute. For RoPE standardization you need to set `self.rope_parameters` in model's config. 
I0618 19:41:14.286735 14977514812544 max_utils.py:259] Attempting to initialize the jax distributed system for GPU backend...
I0618 19:41:14.286977 14977514812544 max_utils.py:261] Jax distributed system initialized on GPU!
I0618 19:41:14.368765 14977514812544 xla_bridge.py:822] Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: libtpu.so: cannot open shared object file: No such file or directory
I0618 19:41:14.592682 14977514812544 pyconfig.py:559] Config param abort_on_inf_loss: True
I0618 19:41:14.592841 14977514812544 pyconfig.py:559] Config param abort_on_nan_loss: True
I0618 19:41:14.592907 14977514812544 pyconfig.py:559] Config param act_quantization_calibration_method: absmax
I0618 19:41:14.592982 14977514812544 pyconfig.py:559] Config param activation_dropout_for_audio: 0.0
I06

CalledProcessError: Command 'b'cd /content/maxtext\n\npython3 -m maxtext.trainers.pre_train.train \\\nsrc/maxtext/configs/base.yml \\\nmodel_name=qwen3-0.6b \\\ndataset_type=synthetic \\\nrun_name=qwen3_gpu_smoke_test \\\nbase_output_directory=/tmp/maxtext_logs \\\nsteps=5 \\\nper_device_batch_size=1 \\\nhardware=gpu \\\nenable_checkpointing=false\n'' returned non-zero exit status 1.

# Task 2 Summary (Current)

| Platform | Status | Reason |
|-----------|---------|---------|
| CPU | Failed | Pallas attention kernel unsupported on CPU backend |
| GPU (Tesla T4) | Failed | Triton requires Ampere GPU (CC ≥ 8.0) |
| TPU | Pending | To be evaluated |

# Qwen 0.6B - TPU Execution

## Objective

Evaluate the execution of the Qwen3-0.6B model on a TPU backend using MaxText.

Unlike the CPU and GPU experiments, TPU is expected to support the kernels required by MaxText and allow successful model initialization and training.

---

## Environment

- Runtime: Google Colab TPU
- MaxText Version: Latest GitHub version
- JAX Version: To be verified
- Flax Version: To be verified

---

## Expected Outcome

The TPU backend should support the required Pallas kernels and avoid the Triton GPU limitations encountered during previous experiments.

The objective is to successfully execute a Qwen3-0.6B synthetic training smoke test.

In [1]:
import jax

print("Devices:")
print(jax.devices())

print("\nBackend:")
print(jax.default_backend())

/usr/local/lib/python3.12/dist-packages/jax/_src/cloud_tpu_init.py:86: UserWarning: Transparent hugepages are not enabled. TPU runtime startup and shutdown time should be significantly improved on TPU v5e and newer. If not already set, you may need to enable transparent hugepages in your VM image (sudo sh -c "echo always > /sys/kernel/mm/transparent_hugepage/enabled")
  warnings.warn(


Devices:
[TpuDevice(id=0, process_index=0, coords=(0,0,0), core_on_chip=0)]

Backend:
tpu


In [2]:
!git clone https://github.com/AI-Hypercomputer/maxtext.git
%cd maxtext

Cloning into 'maxtext'...
remote: Enumerating objects: 97283, done.
remote: Counting objects: 100% (31/31), done.
remote: Compressing objects: 100% (27/27), done.
remote: Total 97283 (delta 7), reused 12 (delta 4), pack-reused 97252 (from 1)
Receiving objects: 100% (97283/97283), 410.33 MiB | 16.91 MiB/s, done.
Resolving deltas: 100% (71474/71474), done.
/content/maxtext


In [3]:
%%bash
cd /content/maxtext
bash src/dependencies/scripts/setup.sh MODE=stable

Checking current directory...
Directory check passed.
Skipping editing needrestart.conf
Checking Python version...
Python version check passed. Continuing with script.
--------------------------------------------------
Get:1 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,632 B]
Get:2 https://cli.github.com/packages stable InRelease [3,917 B]
Get:3 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
Hit:4 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:5 https://r2u.stat.illinois.edu/ubuntu jammy InRelease [6,555 B]
Get:6 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [128 kB]
Get:7 https://cli.github.com/packages stable/main amd64 Packages [354 B]
Get:8 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [127 kB]
Get:9 http://security.ubuntu.com/ubuntu jammy-security/main amd64 Packages [4,026 kB]
Get:10 http://security.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [7,221 kB]
Get:11 http://archive.ubuntu.com/ub

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)


W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)


debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 2.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0  

In [1]:
import jax
import flax

print("Devices:", jax.devices())
print("JAX:", jax.__version__)
print("Flax:", flax.__version__)

/usr/local/lib/python3.12/dist-packages/jax/_src/cloud_tpu_init.py:91: UserWarning: Transparent hugepages are not enabled. TPU runtime startup and shutdown time should be significantly improved on TPU v5e and newer. If not already set, you may need to enable transparent hugepages in your VM image (sudo sh -c "echo always > /sys/kernel/mm/transparent_hugepage/enabled")
  warnings.warn(


Devices: [TpuDevice(id=0, process_index=0, coords=(0,0,0), core_on_chip=0)]
JAX: 0.10.0
Flax: 0.12.7


In [2]:
%%bash
cd /content/maxtext

python3 -m maxtext.trainers.pre_train.train \
src/maxtext/configs/base.yml \
model_name=qwen3-0.6b \
dataset_type=synthetic \
run_name=qwen3_tpu_smoke_test \
base_output_directory=/tmp/maxtext_logs \
steps=5 \
per_device_batch_size=1 \
hardware=tpu \
enable_checkpointing=false

/usr/local/lib/python3.12/dist-packages/jax/_src/cloud_tpu_init.py:91: UserWarning: Transparent hugepages are not enabled. TPU runtime startup and shutdown time should be significantly improved on TPU v5e and newer. If not already set, you may need to enable transparent hugepages in your VM image (sudo sh -c "echo always > /sys/kernel/mm/transparent_hugepage/enabled")
  warnings.warn(
: CommandLine Error: Option 'nvptx-traverse-address-aliasing-limit' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options
bash: line 12:  9673 Aborted                 (core dumped) python3 -m maxtext.trainers.pre_train.train src/maxtext/configs/base.yml model_name=qwen3-0.6b dataset_type=synthetic run_name=qwen3_tpu_smoke_test base_output_directory=/tmp/maxtext_logs steps=5 per_device_batch_size=1 hardware=tpu enable_checkpointing=false


CalledProcessError: Command 'b'cd /content/maxtext\n\npython3 -m maxtext.trainers.pre_train.train \\\nsrc/maxtext/configs/base.yml \\\nmodel_name=qwen3-0.6b \\\ndataset_type=synthetic \\\nrun_name=qwen3_tpu_smoke_test \\\nbase_output_directory=/tmp/maxtext_logs \\\nsteps=5 \\\nper_device_batch_size=1 \\\nhardware=tpu \\\nenable_checkpointing=false\n'' returned non-zero exit status 134.

## Results

- TPU backend detected successfully.
- JAX detected the TPU device.
- MaxText environment installation completed successfully.
- Qwen3-0.6B training initialization was attempted.

Execution failed before model compilation.

---

## Error Encountered

```text
CommandLine Error:
Option 'nvptx-traverse-address-aliasing-limit' registered more than once!

LLVM ERROR:
inconsistency in registered CommandLine options
```

---

## Observations

- TPU hardware was detected correctly.
- The failure occurred before model execution.
- The error originated from the LLVM/XLA runtime environment.
- The issue appears unrelated to Qwen3 configuration or dataset setup.
- The failure is likely caused by dependency conflicts within the Colab TPU environment.

---

## Conclusion

The TPU execution attempt could not proceed due to an LLVM/XLA runtime conflict.

Model configuration loading and TPU detection were successful, but the training process terminated before compilation and execution.

**Status: Failed due to TPU runtime environment issues.**

# Qwen Scaled Model (Run B)

## Objective

Scale the base Qwen3-0.6B model to a larger configuration and evaluate its behavior across CPU, GPU, and TPU backends.

The assignment requires scaling the model beyond the base configuration and documenting the changes made.

In [6]:
%cd /content/maxtext

/content/maxtext


In [7]:
!sed -n '1,100p' src/maxtext/configs/models/qwen3-0.6b.yml

!sed -n '1,100p' src/maxtext/configs/models/qwen3-1.7b.yml

# Copyright 2023–2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# model config for qwen3-0.6b

base_emb_dim: 1024
base_num_query_heads: 16
base_num_kv_heads: 8
base_mlp_dim: 3072
base_num_decoder_layers: 28
head_dim: 128
mlp_activations: ["silu", "linear"] # "hidden_act": "silu" implies SwiGLU
vocab_size: 151936

decoder_block: "qwen3"

normalization_layer_epsilon: 1.0e-6
rope_max_timescale: 1000000

use_qk_norm: true

logits_via_embedding: true # from "tie_word_embeddings": true

## Model Scaling Analysis

To create a larger Dense model, the existing MaxText configuration `qwen3-1.7b` was selected.

### Configuration Changes

| Parameter | Qwen3-0.6B | Qwen3-1.7B |
|------------|------------|------------|
| Embedding Dimension | 1024 | 2048 |
| MLP Dimension | 3072 | 6144 |
| Decoder Layers | 28 | 28 |
| Query Heads | 16 | 16 |
| KV Heads | 8 | 8 |
| Head Dimension | 128 | 128 |

### Why these changes increase model size

The largest increase comes from:

- Doubling the embedding dimension
- Doubling the feed-forward (MLP) dimension

These parameters dominate transformer parameter count, resulting in a much larger model while preserving the overall architecture.

### Why Qwen3-1.7B was chosen

The assignment required scaling the base Qwen3-0.6B model to approximately 1B parameters.

Qwen3-1.7B is already provided by MaxText, represents a substantial increase in capacity, and requires no custom architecture modifications.

# Qwen 1.7B - CPU Execution

In [1]:
%%bash
cd /content/maxtext

python3 -m maxtext.trainers.pre_train.train \
src/maxtext/configs/base.yml \
model_name=qwen3-1.7b \
dataset_type=synthetic \
run_name=qwen3_1_7b_cpu \
steps=50 \
per_device_batch_size=1 \
enable_checkpointing=false

2026-06-18 20:35:09.412342: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
2026-06-18 20:35:11.259179: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2026-06-18 20:35:13.798484: I external/local_xla/xla/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
[transformers] DeepseekV32Config got `key=rope_scaling` in kwargs but hasn't set it as attribute. For RoPE standardization you need to set `self.rope_parameters` in model's config. 
2026-06-18 20:35:49.022901: E external/local_xla/xla/stream_executor/cuda/cuda_platform.cc:51] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)
W0618 20:35:49.210612 14049

CalledProcessError: Command 'b'cd /content/maxtext\n\npython3 -m maxtext.trainers.pre_train.train \\\nsrc/maxtext/configs/base.yml \\\nmodel_name=qwen3-1.7b \\\ndataset_type=synthetic \\\nrun_name=qwen3_1_7b_cpu \\\nsteps=50 \\\nper_device_batch_size=1 \\\nenable_checkpointing=false\n'' returned non-zero exit status 137.

## CPU Result

Unlike the Qwen3-0.6B experiment, which failed due to CPU backend limitations, the Qwen3-1.7B experiment progressed further through initialization.

The process was terminated by the operating system before training began.

### Error

```text
Killed
Exit Status: 137
```

### Analysis

Exit code 137 indicates that the process was terminated by the operating system due to memory pressure.

The larger Qwen3-1.7B configuration requires substantially more memory than the 0.6B model because:

- Embedding dimension increased from 1024 → 2048
- MLP dimension increased from 3072 → 6144

As a result, the Colab CPU runtime was unable to allocate sufficient memory for model initialization.

### Conclusion

CPU execution of the scaled model failed due to memory limitations rather than backend compatibility issues.

In [5]:
%%bash
cd /content/maxtext

python3 -m maxtext.trainers.pre_train.train \
src/maxtext/configs/base.yml \
model_name=qwen3-1.7b \
dataset_type=synthetic \
run_name=qwen3_1_7b_gpu \
steps=50 \
per_device_batch_size=1 \
hardware=gpu \
enable_checkpointing=false

2026-06-18 20:47:36.400879: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
[transformers] DeepseekV32Config got `key=rope_scaling` in kwargs but hasn't set it as attribute. For RoPE standardization you need to set `self.rope_parameters` in model's config. 
W0618 20:48:10.363014 16914649047680 pyconfig.py:113] base_output_directory is not provided; Using local directory called maxtext_output
W0618 20:48:10.446311 16914649047680 pyconfig.py:294] tokenizer_path not found in HF_IDS in maxtext/src/maxtext/utils/globals.py.           Using the default src/maxtext/assets/tokenizers/tokenizer.llama2 instead.           Please pass tokenizer_path in your command if this is not intended.
I0618 20:48:10.446714 16914649047680 max_utils.py:259] Attempting to 

CalledProcessError: Command 'b'cd /content/maxtext\n\npython3 -m maxtext.trainers.pre_train.train \\\nsrc/maxtext/configs/base.yml \\\nmodel_name=qwen3-1.7b \\\ndataset_type=synthetic \\\nrun_name=qwen3_1_7b_gpu \\\nsteps=50 \\\nper_device_batch_size=1 \\\nhardware=gpu \\\nenable_checkpointing=false\n'' returned non-zero exit status 1.

## GPU Result

The Qwen3-1.7B model initialized successfully and progressed further than the CPU execution.

Execution failed during Transformer Engine initialization.

### Error

```text
OSError:
libcudnn_heuristic.so.9: undefined symbol:
_ZTVN5cudnn7backend12OperationSetE
```

### Analysis

The failure occurred while loading Transformer Engine GPU dependencies.

The error indicates an incompatibility between:

- Transformer Engine
- CUDA runtime
- cuDNN libraries available in the Colab environment

The model configuration itself was accepted successfully.

### Conclusion

GPU execution of the scaled model failed because of a dependency/runtime compatibility issue rather than model configuration problems.

The TPU runtime issue observed during the 0.6B experiment originates from the environment before model compilation begins. Since the failure occurs prior to model execution, the same issue is expected to affect the scaled configuration as well.